src/library/scala/collection/immutable/ChampHashMap.scala

```diff
@@ -102,8 +106,288 @@ final class HashMap[K, +V] private[immutable] (private val rootNode: MapNode[K,
+                      }
+
+                      n.updated(rightKey, rightValue, rightOriginalHash, rightImproved, nextShift)
+                    }
+
+                    result.content(newContentSize - compressedNodeIdx - 1) = newNode
+                    compressedNodeIdx += 1
+                    leftNodeIdx += 1
+                    rightDataIdx += 1
+                    result.size += newNode.size
+
+                  } else if ((bitpos & leftDataOnly) != 0) {
+                    result.content(MapNode.TupleLength * compressedDataIdx) =
+                      leftBm.getKey(leftDataIdx).asInstanceOf[AnyRef]
+                    result.content(MapNode.TupleLength * compressedDataIdx + 1) =
+                      leftBm.getValue(leftDataIdx).asInstanceOf[AnyRef]
+                    result.originalHashes(compressedDataIdx) = leftBm.originalHashes(leftDataIdx)
+
+                    compressedDataIdx += 1
+                    leftDataIdx += 1
+                    result.size += 1
+                  } else if ((bitpos & rightDataOnly) != 0) {
+                    result.content(MapNode.TupleLength * compressedDataIdx) =
+                      rightBm.getKey(rightDataIdx).asInstanceOf[AnyRef]
+                    result.content(MapNode.TupleLength * compressedDataIdx + 1) =
+                      rightBm.getValue(rightDataIdx).asInstanceOf[AnyRef]
+                    result.originalHashes(compressedDataIdx) = rightBm.originalHashes(rightDataIdx)
+
+                    compressedDataIdx += 1
+                    rightDataIdx += 1
+                    result.size += 1
+                  } else if ((bitpos & leftNodeOnly) != 0) {
+                    val newNode = leftBm.getNode(leftNodeIdx)
+                    result.content(newContentSize - compressedNodeIdx - 1) = newNode
+                    compressedNodeIdx += 1
+                    leftNodeIdx += 1
+                    result.size += newNode.size
+                  } else if ((bitpos & rightNodeOnly) != 0) {
+                    val newNode = rightBm.getNode(rightNodeIdx)
+                    result.content(newContentSize - compressedNodeIdx - 1) = newNode
+                    compressedNodeIdx += 1
+                    rightNodeIdx += 1
+                    result.size += newNode.size
+                  } else if ((bitpos & leftDataRightDataMigrateToNode) != 0) {
+                    val newNode = {
+                      val leftOriginalHash = leftBm.getHash(leftDataIdx)
+                      val rightOriginalHash = rightBm.getHash(rightDataIdx)
+                      new BitmapIndexedMapNode(0, 0, Array(), Array(), 0)
+                        .updated(leftBm.getKey(leftDataIdx), leftBm.getValue(leftDataIdx), leftOriginalHash, improve(leftOriginalHash), nextShift)
+                        .updated(rightBm.getKey(rightDataIdx), rightBm.getValue(rightDataIdx), rightOriginalHash, improve(rightOriginalHash), nextShift)
```

Comment:
Looks like this could be optimized to create and prefill the content array of the correct size, rather than use `updated` to do the same.

Reply:
Yes, I took the lazy way out haha